### PR TITLE
UI testing

### PR DESCRIPTION
**Pull request by shayne-fletcher**
Loads a sandbox with v5 data and checks that 'Martin' can log in and log out.

- Uses `daml-platform.sh`;
- Tidy `package.json`;
- Better checks for `undefined`;
- Passes `--noEmit` to `tsc` when typechecking tests;
- Reinstates `pretty` and uses it.

### CODE REVIEW

File: daml-platform.sh [hunk 1]
```diff
@@ -35,11 +31,15 @@ NAVIGATOR_PID=$!
 daml json-api --ledger-host $LEDGER_HOST --ledger-port $LEDGER_PORT --http-port $JSON_API_PORT 2>&1 1> $DIR/json-api.log &
 JSON_API_PID=$!
 
-kill_navigator_json_api() {
+# Not a simplification. I have found that only the last of multiple
+# traps runs (and so the sandbox is not terminated) when invoked from
+# the UI test.
```
Comment (hurryabit) [inline daml-platform.sh:36]: That comment only makes sense during the code review. I've acknowledged it. Can you please remove it again? 😃
(In the future, you can simply leave a comment in the GitHub in situations like this.)

File: daml-platform.sh [hunk 2]
```diff
@@ -13,10 +13,6 @@ export DAML_SDK_VERSION=$(cat $DIR/SDK_VERSION)
 
 daml sandbox --port $LEDGER_PORT --ledgerid $LEDGER_ID --wall-clock-time &
 SANDBOX_PID=$!
-kill_sandbox() {
-  kill $SANDBOX_PID || true
-}
-trap kill_sandbox EXIT
```
Comment (hurryabit) [inline daml-platform.sh:19]: We should keep this for the case that the script dies between here and installing the trap further down. Worst case end up with two traps and the second attempt to kill the sandbox won't do any harm.

File: ui/package.json [hunk 3]
```diff
@@ -41,11 +41,16 @@
   },
   "proxy": "http://localhost:7575",
   "devDependencies": {
+    "puppeteer": "^2.1.1",
+    "wait-on": "^4.0.1",
     "@types/jest": "^24.0.18",
-    "@types/jwt-simple": "^0.5.33",
```
Comment (hurryabit) [inline ui/package.json:45]: Why did you move this? It was so nicely sorted. (Which `yarn add` does, btw.)

File: ui/src/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    env: { ...process.env, BROWSER: "none" },
+    stdio: "inherit",
+  });
+  await waitOn({ resources: [`tcp:localhost:${UI_PORT}`] });
+
+  // Launch a browser.
+  browser = await puppeteer.launch();
+}, 40_000);
+
+/* eslint-disable @typescript-eslint/require-await */
```
Comment (hurryabit) [inline ui/src/index.test.tsx:67]: That should not be necessary. There's an `await` in the body.

File: daml-platform.sh (hunk 2 above)
Comment (garyverhaegen-da) [reply daml-platform.sh]: > Worst case end up with two traps and the second attempt to kill the sandbox won't do any harm.

The `trap` command sets the code to execute on signal handling, it does not add to a list of functions or anything smart like that. So you have to be a bit careful if you need more than one cleanup action.
